### DISCUSSION TITLE
How can I retrieve the list of available hardware devices for the BUS?

### DISCUSSION THREAD

**Discussion by Gewerd-Strauss** [SaifAqqad/VMR.ahk#26] [Q&A] [answered]
Hello,

I am currently writing a script using this class - which so far has been quite useful.

I want to connect to a set of devices, but that set differs based on whether or not different bluetooth-devices are connected. 

Additionally, based on the list of available devices, I have to modify the settings for specific apps in the windows audio mixer. While this obviously isn't part of voicemeeter's task, it is part of the required setup - and thus the script needs to know which devices are available to begin with. I wouldn't have to do this if certain apps (looking at you, spotify and firefox) would actually allow me to set their audio-out device - but they don't. 

So far I could not figure out how to retrieve the available devices. 

`VMR().UpdateDevices()` force-updates the list of devices in voicemeeter itself - so that devices connected after voicemeeter was launched then show up in the available devices. But I could not find out how to retrieve that data itself to operate upon on ahk's side.

Is this possible? If not, can you think of any other way of retrieving this list?

Thank you.
Sincerely,
~Gw

**Comment by SaifAqqad** [comment, marked as answer]
You can use [`VMRBus.Devices`](https://github.com/SaifAqqad/VMR.ahk/blob/b96e9c06005fb02cae1f81dfc9dc56b96687f626/src/VMRBus.ahk#L18) which is an array of [`VMRDevice`](https://github.com/SaifAqqad/VMR.ahk/blob/master/src/VMRDevice.ahk) objects.

BTW you don't need to manually call `VMR.UpdateDevices`, as VMR will listen for device connections/disconnections and update the arrays automatically

**Reply by Gewerd-Strauss** [reply]
Thank you :)
